File: 프로그래머스/2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>

using namespace std;

int k = 0;
string answerWord;
int answer = 0;

void backtrack(string s, int cnt) {
    if(s == answerWord) {
        answer = k;
        return;
    }
    if(cnt >= 5) return;
    
    k++;
    backtrack(s+"A", cnt+1);
    k++;
    backtrack(s+"E", cnt+1);
    k++;
    backtrack(s+"I", cnt+1);
    k++;
    backtrack(s+"O", cnt+1);
    k++;
    backtrack(s+"U", cnt+1);
}

int solution(string word) {
    answerWord = word;
    backtrack("", 0);
    return answer;
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
class Solution {
    int ans = 0;
    boolean check = false;
    
    public void dfs(String s, String word) {
        if(check) return;
        if(s.length() > 0 && s.length() <= 5) {
            ans++;
            if(s.equals(word)) {
                check = true;
                return;
            }
        }
        if(s.length() == 5) return;
        
        dfs(s+"A", word);
        dfs(s+"E", word);
        dfs(s+"I", word);
        dfs(s+"O", word);
        dfs(s+"U", word);
    }
    
    public int solution(String word) {
        dfs("", word);
        return ans;
    }
}